app/src/main-process/squirrel-updater.ts

```diff
@@ -1,34 +1,236 @@
-  const exeName = Path.basename(process.execPath)
-
-  const spawnUpdate = function(args: string[]) {
-    try {
-      ChildProcess.spawn(updateDotExe, args, { detached: true })
-    } catch (error) {}
-  }
+const appFolder = Path.resolve(process.execPath, '..')
+const rootAppDir = Path.resolve(appFolder, '..')
+const updateDotExe = Path.resolve(Path.join(rootAppDir, 'Update.exe'))
+const exeName = Path.basename(process.execPath)
+
+// A lot of this code was cargo-culted from our Atom comrades:
+// https://github.com/atom/atom/blob/7c9f39e3f1d05ee423e0093e6b83f042ce11c90a/src/main-process/squirrel-update.coffee.
 
+/**
+ * Handle Squirrel.Windows app lifecycle events.
+ *
+ * Returns a promise which will resolve when the work is done.
+ */
+export function handleSquirrelEvent(eventName: string): Promise<void> | null {
   switch (eventName) {
     case '--squirrel-install':
+      return handleInstalled()
+
     case '--squirrel-updated':
-      spawnUpdate(['--createShortcut', exeName])
-      return true
+      return handleUpdated()
 
     case '--squirrel-uninstall':
-      spawnUpdate(['--removeShortcut', exeName])
-      setTimeout(app.quit, 1000)
-      return true
+      return handleUninstall()
 
     case '--squirrel-obsolete':
-      app.quit()
-      return true
+      return Promise.resolve()
+  }
+
+  return null
+}
+
+async function handleInstalled(): Promise<void> {
+  await createShortcut(['StartMenu', 'Desktop'])
+  await installCLI()
+}
+
+async function handleUpdated(): Promise<void> {
+  await updateShortcut()
+  await installCLI()
+}
+
+async function installCLI(): Promise<void> {
+  const binPath = await writeCLITrampoline()
+  const paths = await getPathSegments()
+  if (paths.indexOf(binPath) < 0) {
+    await setPathSegments([...paths, binPath])
+  }
+}
+
+/**
+ * Get the path for the `bin` directory which exists in our `AppData` but
+ * outside path which includes the installed app version.
+ */
+function getBinPath(): string {
+  const appFolder = Path.resolve(process.execPath, '..')
+  const rootAppDir = Path.resolve(appFolder, '..')
+  return Path.join(rootAppDir, 'bin')
+}
+
+/**
+ * Here's the problem: our app's path contains its version number. So each time
+ * we update, the path to our app changes. So it's Real Hard to add our path
+ * directly to `Path`. We'd have to detect and remove stale entries, etc.
+ *
+ * So instead, we write a trampoline out to a fixed path, still inside our
+ * `AppData` directory but outside the version-specific path. That trampoline
+ * just launches the current version's CLI tool. Then, whenever we update, we
+ * rewrite the trampoline to point to the new, version-specific path. Bingo
+ * bango Bob's your uncle.
+ */
+async function writeCLITrampoline(): Promise<string> {
+  const binPath = getBinPath()
+  const appFolder = Path.resolve(process.execPath, '..')
+  const versionedPath = Path.relative(
+    binPath,
+    Path.join(appFolder, 'resources', 'app', 'static', 'github.bat')
```

Comment:
It seems `atom` went with `.cmd` for their extension and so did `node.js`. I don't know if there's any real difference but perhaps it would be wise to stick to that convention?